Added XFAIL to methods

cekbote opened the pull request:
<!-- Added XFAIL to the two methods. -->

#### References to other Issues or PRs
<!-- "Fixes: #18090" -->


#### Added XFAIL to the methods causing issues. 


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

sylee957 commented:
Release notes should not be added for features that are invisible for users.

Chanakya-Ekbote commented:
> 
> 
> Release notes should not be added for features that are invisible for users.

Thank you. Will make the necessary change.

oscarbenjamin commented on sympy/printing/tests/test_tensorflow.py:
```diff
@@ -129,6 +129,7 @@ def test_tensorflow_printing():
 
 
+@XFAIL
 def test_tensorflow_math():
```
Later on someone might try to remove the XFAIL here if they can see the test passes. There should be a comment indicating why it is marked as XFAIL even though it usually passes. The comment should have a link to the issue.

Chanakya-Ekbote replied:
Ok sure. I'll add that. Thank you.

smichr replied:
Why are we testing random numbers? If we are just trying to see if the tensor flow result is the same as the result when evaluating the expr at the same value then why not just choose a non-pathological value like `-pi` to test for each of them?

oscarbenjamin commented:
The OP should have a reference to the issue that you are addressing. It should not say "fixes" though as this does not fix the issue (it just makes it less annoying).

Chanakya-Ekbote commented:
> 
> 
> The OP should have a reference to the issue that you are addressing. It should not say "fixes" though as this does not fix the issue (it just makes it less annoying).

Sure. Apologies for the inconvenience caused.

sylee957 commented:
I have just extended what the original tests were organized before, after fixing some errors
But it looks like tensorflow precision is poor for matrix inverse or acos, and I’d try to revert everything with selective test cases.

Chanakya-Ekbote commented:
> 
> 
> I have just extended what the original tests were organized before, after fixing some errors
> But it looks like tensorflow precision is poor for matrix inverse or acos, and I’d try to revert everything with selective test cases.

So should I create dedicated tests for matrix inversion or for all the problematic ones?

sylee957 commented:
Yes, it can be better to make matrix inversion test non-random, and also ‘acos’

Chanakya-Ekbote commented:
> 
> 
> Yes, it can be better to make matrix inversion test non-random, and also ‘acos’

So what I noticed in acos was that if the random value gets too high (above a threshold of 0.995) tensorflow and sympy give different values. Hence the max limit I've added for acos is 0.99. I have attached the photo below which shows the discrepancy.
![Capture](https://user-images.githubusercontent.com/51071680/73179434-5d270100-4139-11ea-8380-0815231fce7c.PNG)

I have yet to figure out the issue for the Inverse. It may be the same as before where the limits may have to be changed.

Chanakya-Ekbote commented:
> 
> 
> > Yes, it can be better to make matrix inversion test non-random, and also ‘acos’
> 
> So what I noticed in acos was that if the random value gets too high (above a threshold of 0.995) tensorflow and sympy give different values. Hence the max limit I've added for acos is 0.99. I have attached the photo below which shows the discrepancy.
> ![Capture](https://user-images.githubusercontent.com/51071680/73179434-5d270100-4139-11ea-8380-0815231fce7c.PNG)
> 
> I have yet to figure out the issue for the Inverse. It may be the same as before where the limits may have to be changed.

If I have to write custom tests for the Inverse method, how many should I write?

sylee957 commented:
Only one can be sufficient

Chanakya-Ekbote commented:
> 
> 
> Only one can be sufficient

I have added a test. Could you please le me know if its ready to be merged?

oscarbenjamin commented on sympy/printing/tests/test_tensorflow.py:
```diff
@@ -129,6 +161,8 @@ def test_tensorflow_printing():
 
 
+# The XFAIL is added because of the PR: https://github.com/sympy/sympy/issues/18090
+@XFAIL
```
I think it would be better to write
```
# This (random) test is XFAIL because it fails occasionally
# See https://github.com/sympy/sympy/issues/18090
```

oscarbenjamin replied:
I think it would be better to refer to the issue (#18090) than this PR (#18469)

oscarbenjamin commented:
The OP here still does not reference the related issue

Chanakya-Ekbote commented:
> 
> 
> The OP here still does not reference the related issue

Apologies. Do you mean in the code or in this issue? If its in this issue, then I have addressed to a certain degree, issue #18090 

> 
> 
> The OP here still does not reference the related issue

Aplogies. If you mean in the code, I have added the changes you have mentioned above.
If you mean in this issue, then I have addressed issue #18090 to a certain extent.

sylee957 commented:
I think that the changes with the plot still need to be reverted.
I still worry that things can fail near `0.99`, why not set it like `0.95`?

oscarbenjamin commented:
This is the OP: https://github.com/sympy/sympy/pull/18469#issue-367218195

Chanakya-Ekbote commented:
I apologise. My computer got corrupted, and then while resetting my system, I accidentally deleted the remote branch also (while deleting a few other branches). Can I start a new PR? Apologies once again for the inconvenience caused.

Chanakya-Ekbote commented:
I have reverted the changes in ```plot.py``` and have also changed the comments as well included the value of 0.95 instead of 0.99. Is there something else I should do or can this be merged?

Chanakya-Ekbote commented:
> 
> 
> I have reverted the changes in `plot.py` and have also changed the comments as well as included the value of 0.95 instead of 0.99 in `test_tensorflow.py`. Is there something else I should do or can this be merged?

@sylee957 @oscarbenjamin

oscarbenjamin commented on sympy/printing/tests/test_tensorflow.py:
```diff
@@ -1,5 +1,4 @@
 import random
-
```
Why is this line deleted?

oscarbenjamin commented:
Can you squash this into one commit? Lots of unrelated changes have happened in the previous commits.

The way to squash it is like this:
```
git checkout test # go to this branch
git checkout -b test_backup # make a backup branch!
git checkout test # back to this branch
git log # find the first commit that is not yours and copy the hash (e.g. 1234)
git reset 1234 # The hash you copied
git reset --hard # This deletes everything. Make sure you backed up!
git status
git diff
git commit -a # Make a new commit for the changes
git push --force # This pushes the squashed commits here in the PR.
```

Chanakya-Ekbote commented:
> 
> 
> Can you squash this into one commit? Lots of unrelated changes have happened in the previous commits.
> 
> The way to squash it is like this:
> 
> ```
> git checkout test # go to this branch
> git checkout -b test_backup # make a backup branch!
> git checkout test # back to this branch
> git log # find the first commit that is not yours and copy the hash (e.g. 1234)
> git reset 1234 # The hash you copied
> git reset --hard # This deletes everything. Make sure you backed up!
> git status
> git diff
> git commit -a # Make a new commit for the changes
> git push --force # This pushes the squashed commits here in the PR.
> ```

I apologise for these messy commits. I will make sure not make the same mistake again. I did try to do what you asked for, however, there is an issue. My computer got corrupted in the middle and while doing some cleaning I accidentally deleted my test remote. Although I did start a new branch called test, I have only the logs after Feb 1st. Hence is there some alternative? Apologies once again for this mess. The only code that has been changed is in ```test_tensorflow```. The rest all are just reverts (I was messing around with a few Sympy functions and then forgot about that).

cekbote committed:
Use deterministic tests for tensorflow

There were previously random test failures that only happened occasionally.
This adds a deterministic test and marks the random tests as XFAIL.

oscarbenjamin commented:
I have squashed the commit myself. If you try to push or pull this PR now you will have problems.

Chanakya-Ekbote commented:
> 
> 
> I have squashed the commit myself. If you try to push or pull this PR now you will have problems.

Alright. Thanks. I won't mess with it anymore. Apologies once again for the mess.

oscarbenjamin commented:
Ok, thanks for this. I'll merge it now

Chanakya-Ekbote commented:
> 
> 
> Ok, thanks for this. I'll merge it now

Thank you for all your help. It was fun contributing.
